src/ucs/datastruct/usage_tracker.c

```diff
@@ -0,0 +1,295 @@
+/**
+* Copyright (c) NVIDIA CORPORATION & AFFILIATES, 2023. ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include "usage_tracker.h"
+
+#include <ucs/debug/log.h>
+#include <ucs/debug/memtrack_int.h>
+
+
+ucs_status_t
+ucs_usage_tracker_create(unsigned ticks_per_flush, unsigned active_capacity,
+                         double active_thresh, unsigned eject_thresh,
+                         ucs_usage_tracker_flush_cb_t cb, void *opaque,
+                         ucs_usage_tracker_h *usage_tracker_p)
+{
+    ucs_status_t status;
+    ucs_usage_tracker_h usage_tracker;
+
+    usage_tracker = ucs_calloc(1, sizeof(*usage_tracker), "ucs_usage_tracker");
+    if (usage_tracker == NULL) {
+        ucs_error("failed to allocate Usage Tracker (ticks_per_flush %u, "
+                  "capacity: %u, "
+                  "active_thresh: %.2f, eject_thresh %u)",
+                  ticks_per_flush, active_capacity, active_thresh,
+                  eject_thresh);
+        status = UCS_ERR_NO_MEMORY;
+        goto err;
+    }
+
+    status = ucs_lru_create(active_capacity, &usage_tracker->lru);
+    if (status != UCS_OK) {
+        goto err_free_tracker;
+    }
+
+    kh_init_inplace(usage_tracker_hash, &usage_tracker->hash);
+
+    if (kh_resize(usage_tracker_hash, &usage_tracker->hash,

```

Comment:
why need to resize?

Reply:
to avoid initial resizing by khash

Reply:
why is it a problem?

Reply:
it's an optimization for the regular flow (where number of entries will be the same as active_capacity

Reply:
why does it hurt performance to use regular kh_init_inplace?

Reply:
will be resized during fast path